src/ast/mod.rs

```diff
@@ -1101,405 +978,340 @@ pub enum Action {
-            NormAction::LetVar(symbol, other) => Action::Let(*symbol, Expr::Var(*other)),
-            NormAction::LetLit(symbol, lit) => Action::Let(*symbol, Expr::Lit(lit.clone())),
-            NormAction::Set(NormExpr::Call(head, body), other) => Action::Set(
-                *head,
-                body.iter().map(|s| Expr::Var(*s)).collect(),
-                Expr::Var(*other),
-            ),
-            NormAction::Extract(symbol, variants) => {
-                Action::Extract(Expr::Var(*symbol), Expr::Var(*variants))
-            }
-            NormAction::Delete(NormExpr::Call(symbol, args)) => {
-                Action::Delete(*symbol, args.iter().map(|s| Expr::Var(*s)).collect())
-            }
-            NormAction::Union(lhs, rhs) => Action::Union(Expr::Var(*lhs), Expr::Var(*rhs)),
-            NormAction::Panic(msg) => Action::Panic(msg.clone()),
-        }
+#[allow(clippy::type_complexity)]
+impl<Head, Leaf> GenericActions<Head, Leaf, ()>
+where
+    Head: Clone,
+    Leaf: Clone + Hash + Eq + Clone,
+{
+    pub fn new(actions: Vec<GenericAction<Head, Leaf, ()>>) -> Self {
+        Self(actions)
     }
 
-    pub fn map_exprs(&self, f: &mut impl FnMut(&NormExpr) -> NormExpr) -> NormAction {
-        match self {
-            NormAction::Let(symbol, expr) => NormAction::Let(*symbol, f(expr)),
-            NormAction::LetVar(symbol, other) => NormAction::LetVar(*symbol, *other),
-            NormAction::LetLit(symbol, lit) => NormAction::LetLit(*symbol, lit.clone()),
-            NormAction::Set(expr, other) => NormAction::Set(f(expr), *other),
-            NormAction::Extract(var, variants) => NormAction::Extract(*var, *variants),
-            NormAction::Delete(expr) => NormAction::Delete(f(expr)),
-            NormAction::Union(lhs, rhs) => NormAction::Union(*lhs, *rhs),
-            NormAction::Panic(msg) => NormAction::Panic(msg.clone()),
-        }
+    pub fn singleton(action: GenericAction<Head, Leaf, ()>) -> Self {
+        Self(vec![action])
     }
 
-    // fvar accepts a variable and if it is being defined (true) or used (false)
-    pub(crate) fn map_def_use(&self, fvar: &mut impl FnMut(Symbol, bool) -> Symbol) -> NormAction {
-        match self {
-            NormAction::Let(symbol, expr) => {
-                NormAction::Let(fvar(*symbol, true), expr.map_def_use(fvar, false))
-            }
-            NormAction::LetVar(symbol, other) => {
-                NormAction::LetVar(fvar(*symbol, true), fvar(*other, false))
-            }
-            NormAction::LetLit(symbol, lit) => NormAction::LetLit(fvar(*symbol, true), lit.clone()),
-            NormAction::Set(expr, other) => {
-                NormAction::Set(expr.map_def_use(fvar, false), fvar(*other, false))
-            }
-            NormAction::Extract(var, variants) => {
-                NormAction::Extract(fvar(*var, false), fvar(*variants, false))
+    pub(crate) fn to_norm_actions<FG: FreshGen<Head, Leaf>>(
```

Comment:
we should re-name this and add documentation
- explain why it returns both core actions and generic actions
- explain what the (Head, Leaf) is in the genericactions

Reply:
done